File: app/src/main/java/space/pal/sig/view/launches/LaunchesViewModel.java
package space.pal.sig.view.launches;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.paging.PagedList;

import space.pal.sig.model.Launch;
import space.pal.sig.repository.LaunchesRepository;

/**
 * SpaceNow
 * Created by Catalin on 7/19/2020
 **/
public class LaunchesViewModel extends AndroidViewModel {

    private final LaunchesRepository launchesRepository;
    private final MediatorLiveData<PagedList<Launch>> launchList = new MediatorLiveData<>();
    private final MutableLiveData<Integer> selectedFilter = new MutableLiveData<>();
    private final String[] filterLaunches = new String[]{
            "Future Launches",
            "Past Launches"
    };
    private LiveData<PagedList<Launch>> launchesLiveData;

    public LaunchesViewModel(@NonNull Application application,
                             LaunchesRepository launchesRepository) {
        super(application);
        this.launchesRepository = launchesRepository;
        selectFilter(0);
    }

    public void filter(int pos) {
        if (pos == 0) {
            findFutureLaunches();
        } else {
            findPastLaunches();
        }
    }

    public void selectFilter(int pos) {
        this.selectedFilter.setValue(pos);
    }

    public LiveData<PagedList<Launch>> getLaunchList() {
        return launchList;
    }

    public LiveData<Integer> getSelectedFilter() {
        return selectedFilter;
    }

    public String[] getFilterLaunches() {
        return filterLaunches;
    }

    private void findPastLaunches() {
        launchList.removeSource(launchesLiveData);
        launchesLiveData = launchesRepository.findAllPastLaunches(50);
        launchList.addSource(launchesLiveData, launchList::setValue);
    }

    private void findFutureLaunches() {
        launchList.removeSource(launchesLiveData);
        launchesLiveData = launchesRepository.findAllFutureLaunches(50);
        launchList.addSource(launchesLiveData, launchList::setValue);
    }
}

File: app/src/main/java/space/pal/sig/view/launches/LaunchesViewModel.kt
package space.pal.sig.view.launches

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import space.pal.sig.R
import space.pal.sig.model.entity.LaunchWithData
import space.pal.sig.repository.LaunchRepository
import space.pal.sig.view.BaseViewModel

/**
 * SpaceNow
 * Created by Catalin on 12/23/2020
 **/
class LaunchesViewModel(
        private val launchRepository: LaunchRepository,
        application: Application
) : BaseViewModel(application) {

    private val launches = MediatorLiveData<MutableList<LaunchWithData>>()
    private val selectedOption = MutableLiveData<String>()
    private val options = arrayOf(
            application.getString(R.string.upcoming_launches),
            application.getString(R.string.past_launches),
    )
    private var launchesLiveData: LiveData<MutableList<LaunchWithData>>? = null

    init {
        filter(0)
    }

    fun getLaunches(): LiveData<MutableList<LaunchWithData>> {
        return launches
    }

    fun getSelectedOption(): LiveData<String> {
        return selectedOption
    }

    fun getOptions(): Array<String> {
        return options
    }

    fun filter(pos: Int) {
        when (pos) {
            0 -> {
                findFutureLaunches()
            }
            1 -> {
                findPastLaunches()
            }
        }
        selectedOption.value = options[pos]
    }

    private fun findFutureLaunches() {
        launchesLiveData?.let { launches.removeSource(it) }
        launchesLiveData = launchRepository.findAllFromFuture()
        launchesLiveData?.let {
            launches.addSource(it) { data ->
                launches.value = data
            }
        }
    }

    private fun findPastLaunches() {
        launchesLiveData?.let { launches.removeSource(it) }
        launchesLiveData = launchRepository.findAllFromPast()
        launchesLiveData?.let {
            launches.addSource(it) { data ->
                launches.value = data
            }
        }
    }

}